apphub/semantic_segmentation/unet3d/3DUnetMRBrainSegmentation_Isensee.py

```diff
@@ -0,0 +1,228 @@
+                                          lambda:self.handle_nopaddding_case(patch))
+        return data_patch, truth_patch
+
+    def forward(self, data, state):
+        self.modality_images, self.seg_mask = data
+        cls_1 = tf.where(self.seg_mask[0] == self.labels[0], 1, 0)
+        cls_2 = tf.where(self.seg_mask[0] == self.labels[1], 1, 0)
+        cls_4 = tf.where(self.seg_mask[0] == self.labels[2], 1, 0)
+        self.segmask_multi_class = tf.stack([cls_1, cls_2, cls_4])
+
+        # We randomly select the start of the patches
+        random_start_offset = tf.negative(
+            tf.random.uniform(shape=(3, ), minval=0, maxval=self.patch_start_offset + 1, dtype=tf.dtypes.int32))
+
+        start = random_start_offset
+        stop = self.image_shape + random_start_offset
+        step = self.patch_shape
+        x,y,z = tf.meshgrid( tf.range(start[0],stop[0],step[0]),  tf.range(start[1],stop[1],step[1]),
+                                        tf.range(start[2],stop[2],step[2]),indexing='ij')
+        patches = tf.stack([x, y, z], axis=0)
+        patches = tf.transpose(tf.reshape(patches, (3, -1)))
+
+        data_stack,truth_stack = tf.map_fn(self.get_image_patch,patches,
+                                           dtype=(tf.dtypes.float32, tf.dtypes.int32), back_prop=False)
+        mask = tf.map_fn(lambda tr: tf.reduce_any(tf.not_equal(tr, 0)), truth_stack, dtype=tf.bool, back_prop=False)
+        data_stack = tf.boolean_mask(data_stack, mask)
+        truth_stack = tf.boolean_mask(truth_stack, mask)
+
+        return data_stack, truth_stack
+
+
+class SplitMaskLabelwise(TensorOp):
+    def __init__(self, inputs=None, outputs=None, mode=None):
+        super().__init__(inputs=inputs, outputs=outputs, mode=mode)
+        self.nlabels = 3
+        self.labels = [1, 2, 4]
+
+    def forward(self, data, state):
+        seg_mask = data
+        cls_1 = tf.where(seg_mask[0] == self.labels[0], 1, 0)
+        cls_2 = tf.where(seg_mask[0] == self.labels[1], 1, 0)
+        cls_4 = tf.where(seg_mask[0] == self.labels[2], 1, 0)
+        segmask_multi_label = tf.stack([cls_1, cls_2, cls_4])
+        return segmask_multi_label
+
+
+class WeightedDiceLoss(Loss):
+    def forward(self, data, state):
+        y_true, y_pred = data
+        smooth = 0.00001
+        y_true = tf.cast(y_true, dtype=tf.dtypes.float32)
+        intersection = tf.reduce_sum(y_true * y_pred, axis=[-3, -2, -1])
+        union = tf.reduce_sum(y_true, axis=[-3, -2, -1]) + tf.reduce_sum(y_pred, axis=[-3, -2, -1])
+        loss = -2. * (intersection + smooth / 2) / (union + smooth)
+        loss = tf.reduce_mean(loss, axis=1)
+        return loss
+
+
+def get_estimator(batch_size=1, epochs=500, steps_per_epoch=128, model_dir=tempfile.mkdtemp(), path_brats=None):
+    """Args:
+        path_brats: folder path of BraTs 2018 dataset, containing data subdir inturn having LGG and HGG data.
```

Comment:
why are we not expecting them to put the dataset inside the ~/fastestimator_data folder like we do all of the other datasets?

Reply:
To download BraTs 2018 data , there are few steps:
# register at their site  and request for data
# receive a mail from them , follow the link in the mail's attachment to download the data.
# the links in the attachments are valid only for a certain time
Programming this part , a challenge.

Reply:
that's fine, but we should still ask them to move it to the ~/fastestimator_data folder rather than expecting it in a random folder

Reply:
fastestimator/dataset/brats.py has logic to ensure that  input  BraTs folder has to be subfolder similar to ~/fastestimator_data/BraTs.

Reply:
in that case lets make the default arg to get_esimator be "~/fastestimator_data/BraTs" since we're asserting that the path is not None on line 144

Reply:
The default arg to get_estimator is now os.path.join(os.getenv('HOME'), 'fastestimator_data', 'BraTs')